lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,112 @@
-  // Make sure to set these to more appropriate limits for
-  // your application.
-  setInputLimits(0.0, 3.3);
-  setOutputLimits(0.0, 3.3);
-
-  tSample_ = interval;
-
-  setTunings(Kc, tauI, tauD);
-
-  setPoint_             = 0.0;
-  processVariable_      = 0.0;
-  prevProcessVariable_  = 0.0;
-  controllerOutput_     = 0.0;
-  prevControllerOutput_ = 0.0;
-
-  accError_      = 0.0;
-  deadZoneError_ = 0.0;
-  bias_          = 0.0;
+PID::PID(uint32_t proportionalGain, uint32_t intregralGain, uint32_t derivativeGain, int32_t lowerBound,
+         int32_t upperBound, float deadzone, bool antiKickback)
+    : m_PGain(proportionalGain),
+      m_IGain(intregralGain),
+      m_DGain(derivativeGain),
+      m_lowerBound(lowerBound),
+      m_upperBound(upperBound),
+      m_deadzone(deadzone),
+      m_antiKickback(antiKickback),
+      m_IPath(0),
+      m_pastError(0),
+      m_pastPV(0) {}
 
-  realOutput_ = 0.0;
+mbed_error_status_t PID::updateProportionalGain(uint32_t p) {
+  m_PGain.store(p);
+  return MBED_SUCCESS;
 }
 
-void PID::setInputLimits(float inMin, float inMax) {
-  // Make sure we haven't been given impossible values.
-  if (inMin >= inMax) {
-    return;
-  }
-
-  inSpan_ = inMax - inMin;
-
-  // Rescale the working variables to reflect the changes.
-  prevProcessVariable_ *= (inMax - inMin) / inSpan_;
-  accError_ *= (inMax - inMin) / inSpan_;
-
-  // Make sure the working variables are within the new limits.
-  if (prevProcessVariable_ > 1) {
-    prevProcessVariable_ = 1;
-  } else if (prevProcessVariable_ < 0) {
-    prevProcessVariable_ = 0;
-  }
-
-  inMin_ = inMin;
-  inMax_ = inMax;
-}
-
-void PID::setOutputLimits(float outMin, float outMax) {
-  // Make sure we haven't been given impossible values.
-  if (outMin >= outMax) {
-    return;
-  }
-
-  // Rescale the working variables to reflect the changes.
-  prevControllerOutput_ *= (outMax - outMin) / outSpan_;
-
-  // Make sure the working variables are within the new limits.
-  if (prevControllerOutput_ > 1) {
-    prevControllerOutput_ = 1;
-  } else if (prevControllerOutput_ < 0) {
-    prevControllerOutput_ = 0;
-  }
-
-  outMin_  = outMin;
-  outMax_  = outMax;
-  outSpan_ = outMax - outMin;
-}
-
-void PID::setTunings(float Kc, float tauI, float tauD) {
-  // Verify that the tunings make sense.
-  if (Kc == 0.0 || tauI < 0.0 || tauD < 0.0) {
-    return;
-  }
-
-  // Store raw values to hand back to user on request.
-  pParam_ = Kc;
-  iParam_ = tauI;
-  dParam_ = tauD;
-
-  float tempTauR;
-
-  if (tauI == 0.0) {
-    tempTauR = 0.0;
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+mbed_error_status_t PID::updateIntegralGain(uint32_t i) {
```

Comment:
Do we need any handling for if something goes wrong and it shouldn't return MBED_SUCCESS? Otherwise, is it even useful to return values at all if it's only going to return MBED_SUCCESS?

Reply:
yeah this will likely change. im still trying to figure out if we're gonna lock a mutex in these functions. but yeah definetly wont just return success always